src/GraphBLAS-sharp.Backend/Objects/Vector.fs

```diff
@@ -0,0 +1,188 @@
+﻿namespace GraphBLAS.FSharp.Backend
+
+open Brahma.FSharp
+open GraphBLAS.FSharp.Backend
+
+type VectorFormat =
+    | COO
+    | Dense
+
+type COOVector<'a> =
+    { Size: int
+      Indices: int []
+      Values: 'a [] }
+
+    override this.ToString() =
+        [ sprintf "Sparse Vector\n"
+          sprintf "Size:    %i \n" this.Size
+          sprintf "Indices: %A \n" this.Indices
+          sprintf "Values:  %A \n" this.Values ]
+        |> String.concat ""
+
+    member this.ToDevice(context: ClContext) =
+        let indices = context.CreateClArray this.Indices
+        let values = context.CreateClArray this.Values
+
+        { Context = context
+          Size = this.Size
+          Indices = indices
+          Values = values }
+
+    static member FromTuples(size: int, indices: int [], values: 'a []) =
+        { Size = size
+          Indices = indices
+          Values = values }
+
+    static member FromArray(array: 'a [], isZero: 'a -> bool) =
+        let (indices, vals) =
+            array
+            |> Seq.cast<'a>
+            |> Seq.mapi (fun idx v -> (idx, v))
+            |> Seq.filter (fun (_, v) -> not (isZero v))
+            |> Array.ofSeq
+            |> Array.unzip
+
+        COOVector.FromTuples(array.Length, indices, vals)
+
+and ClCooVector<'a> =
+    { Context: ClContext
+      Size: int
+      Indices: ClArray<int>
+      Values: ClArray<'a> }
+
+    member this.ToHost(q: MailboxProcessor<_>) =
+        let indices = Array.zeroCreate this.Indices.Length
+        let values = Array.zeroCreate this.Values.Length
+
+        let _ =
+            q.Post(Msg.CreateToHostMsg(this.Indices, indices))
+
+        let _ =
+            q.PostAndReply(fun ch -> Msg.CreateToHostMsg(this.Values, values, ch))
+
+        { Size = this.Size
+          Indices = indices
+          Values = values }
+
+    interface IDeviceMemObject with
+        member this.Dispose(q) =
+            q.Post(Msg.CreateFreeMsg<_>(this.Values))
+            q.Post(Msg.CreateFreeMsg<_>(this.Indices))
+            q.PostAndReply(Msg.MsgNotifyMe)
+
+    member this.Dispose(q) = (this :> IDeviceMemObject).Dispose(q)
+
+type DenseVector<'a> =
```

Comment:
What the differrence from ```ClArray```? Can we unify these types? Or make ```DenceVector``` (or ```clDenceVector```) an extension of ```ClArray```?